File: js/nebulachain/tests/services/notifications.service.test.js
const test = require('node:test');
const assert = require('node:assert/strict');
const notifications = require('../../services/notifications/service');

test('NotificationPlanner plans channels by severity', () => {
  const planner = new notifications.NotificationPlanner();
  const plan = planner.plan('op1', 5);
  assert.ok(plan.channels.includes('pager'));
});

test('shouldThrottle enforces window limit', () => {
  assert.equal(notifications.shouldThrottle({ recentCount: 11, maxPerWindow: 10, severity: 3 }), true);
  assert.equal(notifications.shouldThrottle({ recentCount: 5, maxPerWindow: 10, severity: 3 }), false);
});

test('formatNotification includes severity tag', () => {
  const n = notifications.formatNotification({ operatorId: 'op1', severity: 6, message: 'alert' });
  assert.ok(n.body.includes('[SEV-6]'));
});

test('batchNotify sends to all operators', () => {
  const result = notifications.batchNotify({ operators: ['a', 'b', 'c'], severity: 4, message: 'test' });
  assert.equal(result.length, 3);
});

File: ruby/opalcommand/tests/services/notifications_service_test.rb
# frozen_string_literal: true

require_relative '../test_helper'
require_relative '../../services/notifications/service'

class NotificationsServiceTest < Minitest::Test
  def test_notification_planner_plan_and_pending
    planner = OpalCommand::Services::Notifications::NotificationPlanner.new
    planner.plan(operator_id: 'op1', severity: 3, message: 'test alert')
    assert_equal 1, planner.size
    assert_equal 1, planner.pending.length
  end

  def test_should_throttle_at_limit
    assert OpalCommand::Services::Notifications.should_throttle(recent_count: 10, max_per_window: 10)
    refute OpalCommand::Services::Notifications.should_throttle(recent_count: 5, max_per_window: 10)
  end

  def test_format_notification_includes_channels
    notif = OpalCommand::Services::Notifications.format_notification(operator_id: 'op1', severity: 4, message: 'urgent')
    assert_includes notif[:channels], 'pager'
    assert_equal 4, notif[:severity]
  end

  def test_notification_summary_groups_by_severity
    batch = [
      { severity: 1, message: 'a' },
      { severity: 1, message: 'b' },
      { severity: 3, message: 'c' }
    ]
    summary = OpalCommand::Services::Notifications.notification_summary(batch)
    assert_equal 3, summary[:total]
    assert_equal 2, summary[:by_severity][1]
  end
end
